snuba/datasets/plans/storage_plan_builder_new.py

```diff
@@ -0,0 +1,290 @@
+from snuba.pipeline.utils.storage_finder import StorageKeyFinder
+from snuba.query import Query as AbstractQuery
+from snuba.query.allocation_policies import AllocationPolicy
+from snuba.query.data_source.simple import Table
+from snuba.query.logical import Query as LogicalQuery
+from snuba.query.processors.physical import ClickhouseQueryProcessor
+from snuba.query.processors.physical.conditions_enforcer import (
+    MandatoryConditionEnforcer,
+)
+from snuba.query.processors.physical.mandatory_condition_applier import (
+    MandatoryConditionApplier,
+)
+from snuba.query.query_settings import QuerySettings
+from snuba.state import explain_meta
+from snuba.utils.metrics.util import with_span
+
+TQuery = TypeVar("TQuery", bound=AbstractQuery)
+
+
+@dataclass(frozen=True)
+class QueryPlanNew(ABC, Generic[TQuery]):
+    """
+    Provides the directions to execute a Clickhouse Query against one
+    storage or multiple joined ones.
+
+    This is produced in the storage processing stage of the query pipeline.
+
+    It embeds the Clickhouse Query (the query to run on the storage
+    after translation). It also provides a plan execution strategy
+    that takes care of coordinating the execution of the query against
+    the database.
+
+    When running a query we need a cluster, the cluster is picked according
+    to the storages sets containing the storages used in the query.
+    So the plan keeps track of the storage set as well.
+    There must be only one storage set per query.
+    """
+
+    query: TQuery
+    storage_set_key: StorageSetKey
+
+
+@dataclass(frozen=True)
+class ClickhouseQueryPlanNew(QueryPlanNew[Query]):
+    """
+    Query plan for a single entity, single storage query.
+
+    It provides the sequence of storage specific QueryProcessors
+    to apply to the query after the the storage has been selected.
+    These are divided in two sequences: plan processors and DB
+    processors.
+    Plan processors and DB Query Processors are both executed only
+    once per plan.
+    """
+
+    # Per https://github.com/python/mypy/issues/10039, this has to be redeclared
+    # to avoid a mypy error.
+    plan_query_processors: Sequence[ClickhouseQueryProcessor]
+    db_query_processors: Sequence[ClickhouseQueryProcessor]
+
+
+def get_query_data_source(
+    relational_source: RelationalSource,
+    allocation_policies: list[AllocationPolicy],
+    final: bool,
+    sampling_rate: Optional[float],
+    storage_key: StorageKey,
+) -> Table:
+    assert isinstance(relational_source, TableSource)
+    return Table(
+        table_name=relational_source.get_table_name(),
+        schema=relational_source.get_columns(),
+        allocation_policies=allocation_policies,
+        final=final,
+        sampling_rate=sampling_rate,
+        mandatory_conditions=relational_source.get_mandatory_conditions(),
+        storage_key=storage_key,
+    )
+
+
+class EntityProcessingExecutor:
```

Comment:
should this class be in an entity-related file in order to separate it more from the storage-related concepts?

Reply:
Yes, I agree! After figuring out the composite query pipeline, I plan to rename this file to something more appropriate. By keeping all the new classes here and with a filename suffixed with `-new` helps me differentiate between code that supports old and new pipeline.